packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,549 @@
-      })
-    }
+   * When a websocket is closed by something else - reconnect
+   * @private
+   */
+  onWebSocketClose(event) {
+    logger.info('The current websocket was closed by a third party', event)
+    this.retryManager.onFailure(event)
+    this.reconnect()
+  }
 
-    return new Promise(resolve => {
-      const key = generateKey(eventName, type, id)
-      this.on(key, handler)
+  /* * * * * * * * * * * * * */
+  /* * Cozy Client EVENTS  * */
+  /* * * * * * * * * * * * * */
 
-      this._socket.once(`subscribe_${type}_${id}`, resolve)
-      this._socket.subscribe(type, id)
-    })
+  /**
+   * Register event handlers for cozy-client events
+   * @private
+   */
+  subscribeClientEvents() {
+    this.client.on('login', this.onClientLogin)
+    this.client.on('tokenRefreshed', this.onClientTokenRefreshed)
+    this.client.on('logout', this.onClientLogout)
   }
 
   /**
-   * Remove the given handler from the list of handlers for given
-   * doctype/document and event.
-   *
-   * @param {String}  type      Document doctype to unsubscribe from
-   * @param {String}  id        Document id to unsubscribe from
-   * @param {String}  eventName Event to unsubscribe from
-   * @param {Function}  handler   Function to call when an event of the
-   * given type on the given doctype or document is received from stack.
+   * Unregister event handlers for cozy-client events
+   * @private
    */
-  unsubscribe(eventName, type, handlerOrId, handlerOrUndefined) {
-    const { handler, id } = getHandlerAndId(handlerOrId, handlerOrUndefined)
-    validateParameters(eventName, type, id, handler)
-    const key = generateKey(eventName, type, id)
-
-    this.removeListener(key, handler)
+  unsubscribeClientEvents() {
+    this.client.removeListener('login', this.onClientLogin)
+    this.client.removeListener('tokenRefreshed', this.onClientTokenRefreshed)
+    this.client.removeListener('logout', this.onClientLogout)
+  }
 
-    if (!this._haveEventHandler()) {
-      this._resetSocket()
+  /**
+   * When the cozy-client instance successfully login - re-authenticate
+   * @private
+   */
+  onClientLogin() {
+    logger.info('Received login from cozy-client')
+    if (this.isWebSocketOpen()) {
+      this.authenticate()
+      // send subscriptions again, permissions may have changed
+      this.sendSubscriptions()
     }
   }
 
   /**
-   * Unsubscribe all handlers and close socket
+   * When the cozy-client instance refresh its access token - re-authenticate
+   * @private
    */
-  unsubscribeAll() {
-    this._getEventKeys()
-      .map(key => this._events[key].length > 0 && key)
-      .filter(Boolean)
-      .forEach(key => (this._events[key] = []))
+  onClientTokenRefreshed() {
+    // In theory, we don't have to do anything, and there is no need to authenticate again.
+    // We are reconnecting anyways because if we have a really bad timing, we may have sent an
+    // expired token for authentication, and we need to reconnect if that happen
```

Comment:
I don't understand the "really bad timing" part here. I think the "In theory we don't have to do anything" can be removed if we know that we a "really bad timing" is possible.

Reply:
:+1: